subscriptions/src/subscriptions.ts

```diff
@@ -0,0 +1,214 @@
+import * as im from 'immutable';
+import { from, merge, Observable, Observer, Subject } from 'rxjs';
+import {
+  startWith,
+  concatMap,
+  flatMap,
+  filter,
+  finalize,
+  ignoreElements,
+  scan,
+  tap,
+  map,
+  publishReplay,
+  refCount,
+  share,
+} from 'rxjs/operators';
+import { applyReducer, Operation } from 'fast-json-patch';
+import {
+  KeyedPatches,
+  Patch,
+  Version,
+  Versioned,
+} from '@binaris/shift-interfaces/dist/subscriptions';
+import { mapWithState, StateAndOutput, takeUntilLast } from './rxutils';
+
+interface SubscriptionState<T> extends Versioned<T> {
+  readonly patches?: ReadonlyArray<Patch>;
+}
+
+type VersionedStateGetter<T> = (key: string) => Promise<Versioned<T>>;
+
+export const REGISTER: 'register' = 'register';
+export const DEREGISTER: 'deregister' = 'deregister';
+const UPDATE: 'update' = 'update';
+
+interface RegisterPollerInput {
+  readonly action: 'register';
+  readonly key: string;
+  readonly version: Version;
+}
+interface DeregisterPollerInput {
+  readonly action: 'deregister';
+  readonly key: string;
+}
+interface UpdatePollerInput {
+  readonly action: 'update';
+  readonly keysToVersions: Array<[string, Version]>;
+}
+type PollerInput = RegisterPollerInput | DeregisterPollerInput;
+
+export type NonEmptyArray<T> = [T, ...T[]];
+
+export function isNonEmptyArray<T>(arr: ReadonlyArray<T>): arr is NonEmptyArray<T> {
```

Comment:
This is `const`-unsafe (if you'll excuse my C++): it transforms nonempty `ReadonlyArray`s into `Array`s.  AFAICT the parameter has to be `T[]`, or you need to define
```ts
export type NonEmptyReadonlyArray<T> = NonEmptyArray<T> & ReadonlyArray<T>
```
and use that (untested).